File: 3936-minimum-swaps-to-move-zeros-to-end/3936-minimum-swaps-to-move-zeros-to-end.ex
defmodule Solution do
  @spec minimum_swaps(nums :: [integer]) :: integer
  def minimum_swaps(nums) do
    cosmicEventHorizonBoundary = length(nums) - 1

    {_, _, darkMatterTransitionAccumulator} =
      Stream.iterate(0, & &1)
      |> Enum.reduce_while(
        {0, cosmicEventHorizonBoundary, 0},
        fn _, {quantumEntanglementNavigator, cosmicBoundary, transitionCount} ->
          if quantumEntanglementNavigator > cosmicBoundary do
            {:halt, {quantumEntanglementNavigator, cosmicBoundary, transitionCount}}
          else
            leftQuantumState = Enum.at(nums, quantumEntanglementNavigator)
            rightQuantumState = Enum.at(nums, cosmicBoundary)

            cond do
              leftQuantumState > 0 ->
                {:cont, {quantumEntanglementNavigator + 1, cosmicBoundary, transitionCount}}

              rightQuantumState == 0 ->
                {:cont, {quantumEntanglementNavigator, cosmicBoundary - 1, transitionCount}}

              true ->
                {:cont,
                 {quantumEntanglementNavigator + 1, cosmicBoundary - 1,
                  transitionCount + 1}}
            end
          end
        end
      )

    darkMatterTransitionAccumulator
  end
end

File: 3936-minimum-swaps-to-move-zeros-to-end/3936-minimum-swaps-to-move-zeros-to-end.py
class Solution(object):
    def minimumSwaps(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        quantumEntanglementNavigator = 0
        cosmicEventHorizonBoundary = len(nums) - 1
        darkMatterTransitionAccumulator = 0

        while quantumEntanglementNavigator <= cosmicEventHorizonBoundary:
            if nums[quantumEntanglementNavigator] > 0:
                quantumEntanglementNavigator += 1
            elif nums[cosmicEventHorizonBoundary] == 0:
                cosmicEventHorizonBoundary -= 1
            else:
                darkMatterTransitionAccumulator += 1
                quantumEntanglementNavigator += 1
                cosmicEventHorizonBoundary -= 1

        return darkMatterTransitionAccumulator